### PR TITLE
Wr 308141

### PR DESCRIPTION
**Pull request by kristian-94**

### CODE REVIEW

File: classes/plugininfo/cleaner.php [hunk 1]
```diff
@@ -148,6 +148,10 @@ public function get_settings_section_url() {
             return new \moodle_url('/local/datacleaner/cleaner/environment_matrix/index.php');
         }
 
+        if ($this->name == 'scheduled_tasks') {
+            return new \moodle_url('/local/datacleaner/cleaner/scheduled_tasks/index.php');
```
Comment (brendanheywood) [inline classes/plugininfo/cleaner.php]: I know there is a precedent with the environment_matrix (which should also be fixed), but this should be overridden in the plugin, not in the base class. The base should never be aware of the different cleaners

File: cleaner/scheduled_tasks/classes/clean.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        } else {
+            if ($dryrun) {
+                mtrace("Would disable the {$count} following tasks:");
+            } else {
+                mtrace("Disabling the {$count} following tasks:");
+            }
+
+            // Disable every task that has a record in our table.
+            foreach ($disabledtasks as $disabledtask) {
+                foreach ($tasknames as $taskname) {
```
Comment (dkleto) [inline cleaner/scheduled_tasks/classes/clean.php]: We could avoid iterating through all of the scheduled tasks here by using get_records_select_menu to get an associative array of tasks first and doing a lookup on that, or alternatively pushing this work to the DB by writing the two queries above as one sql statement with a join.

File: cleaner/scheduled_tasks/classes/form/task_form.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+class task_form extends moodleform {
+    protected function definition() {
+        $mform = $this->_form;
+        $tasks = $this->_customdata;
+
+        if (!$tasks) {
+            throw new \coding_exception(get_string('noscheduledtasks', 'cleaner_scheduled_tasks'));
+        }
+        get_string('pluginname', 'cleaner_scheduled_tasks');
```
Comment (dkleto) [inline cleaner/scheduled_tasks/classes/form/task_form.php]: Should the output of the call to get_string be assigned to anything here?

File: cleaner/scheduled_tasks/classes/form/task_form.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        // Display a header on the page.
+        // Need to put lang strings into lang folder later.
```
Comment (dkleto) [inline cleaner/scheduled_tasks/classes/form/task_form.php]: Looks like you've already set up a lang strings file so this comment could be removed.

File: cleaner/scheduled_tasks/classes/form/task_form.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            $mform->setDefault($cbkey, "$default");
+
+            $nexttaskcomponent = (isset($tasks[$key + 1])) ? $tasks[$key + 1]->get_component() : null;
+
+            if (isset($addcomponent[0]) && $component != $nexttaskcomponent) {
+                // if different component to the next, then we add group
+                $mform->addGroup($rendertasks, "$class", "$component", array(' '), false);
+                $rendertasks = [];
+            }
+            $i++;
```
Comment (dkleto) [inline cleaner/scheduled_tasks/classes/form/task_form.php]: I don't see $i being used for anything - can this be removed?

File: cleaner/scheduled_tasks/db/install.xml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<XMLDB PATH="local/datacleaner/cleaner/scheduled_tasks/db" VERSION="2019030400" COMMENT="XMLDB file for Moodle local/datacleaner/cleaner/scheduled_tasks"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="../../../../../lib/xmldb/xmldb.xsd"
+>
+    <TABLES>
+        <TABLE NAME="cleaner_scheduled_tasks" COMMENT="Scheduled tasks config - whether to disable the task in a clean">
+            <FIELDS>
+                <FIELD NAME="id" TYPE="int" LENGTH="10" NOTNULL="true" SEQUENCE="true"/>
+                <FIELD NAME="task_scheduled_id" TYPE="int" LENGTH="10" NOTNULL="true" SEQUENCE="false"/>
```
Comment (dkleto) [inline cleaner/scheduled_tasks/db/install.xml]: I can't actually find it anywhere in the coding style guide, but table fields elsewhere in Moodle don't use underscores, so let's stick with this convention.

File: cleaner/scheduled_tasks/index.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    // If we submit the form, then we should look at the data here and for each record insert the data into our cleaner_scheduled_tasks table.
+    global $DB;
+
+    $taskdata = isset($data->selected) ? $data->selected : false;
+    $taskdata = $taskdata ? $taskdata : (array)$data;
+
+    $scheduledtasks = $DB->get_records('task_scheduled');
+
+    foreach ($taskdata as $key => $taskenabled) {
+        foreach ($scheduledtasks as $scheduledtask) {
```
Comment (dkleto) [inline cleaner/scheduled_tasks/index.php]: This is another case where we could avoid iterating through the full list of scheduled tasks once for every task we're processing (like in the "execute" method above).

File: cleaner/scheduled_tasks/classes/clean.php (hunk 2 above)
Comment (brendanheywood) [reply cleaner/scheduled_tasks/classes/clean.php]: I'd do this in a single update sql, just update all where $DB->get_in_or_equal( list of tasks)

File: cleaner/scheduled_tasks/classes/form/task_form.php [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        $headerarray = [];
+        $headerarray[] = &$mform->createElement('static', 'stitle', 'stitle', "$header $headersubtitle");
+        $mform->addGroup($headerarray, 'header_array', '' , ' ', false);
+
+        // Header for each column
+        $mform->addElement('static', 'description', 'Component', 'Task Name');
+
+        global $DB;
+        $cleanertasks = $DB->get_records_sql("select * from {cleaner_scheduled_tasks} cs
+												join {task_scheduled} ts on ts.id = cs.task_scheduled_id");
```
Comment (brendanheywood) [inline cleaner/scheduled_tasks/classes/form/task_form.php]: whitespace ?

Also try to keep all the sql keywords in caps

https://docs.moodle.org/dev/SQL_coding_style#Parameter_placeholders

File: cleaner/scheduled_tasks/db/install.xml (hunk 6 above)
Comment (brendanheywood) [reply cleaner/scheduled_tasks/db/install.xml]: docs are here:

https://docs.moodle.org/dev/XMLDB_defining_an_XML_structure#Conventions

File: classes/plugininfo/cleaner.php [hunk 9]
```diff
@@ -141,11 +141,14 @@ public function get_settings_section_name() {
...
-        if ($this->name == 'environment_matrix') {
-            return new \moodle_url('/local/datacleaner/cleaner/environment_matrix/index.php');
+        $sectionname = $this->get_settings_section_name();
+        $classname = "\\" . $sectionname . "\\clean";
+
+        if ($sectionname && method_exists($classname, 'get_settings_section_url')) {
+            return $classname::get_settings_section_url($sectionname);
         }
 
         if (file_exists($this->full_path('settings.php'))) {
```
Comment (brendanheywood) [inline classes/plugininfo/cleaner.php]: So this code should never execute?

File: cleaner/scheduled_tasks/classes/clean.php [hunk 10]
```diff
@@ -56,27 +59,29 @@ static public function execute() {
...
+                        $taskstoupdate[] = $updatetask;
+                        //$DB->update_record('task_scheduled', $updatetask);
+                        $increment++;
                     }
                 }
             }
+            if ($taskstoupdate) {
+                // Collect all the tasks to update and update in one query
+                list($sql, $params) = $DB->get_in_or_equal($taskstoupdate);
+                $DB->execute("UPDATE {task_scheduled} SET disabled = 1 WHERE id " . $sql, $params);
```
Comment (kristian-94) [inline cleaner/scheduled_tasks/classes/clean.php]: @dkleto This part wasn't working for me, which is why it's been reverted in the current version. The code expects id to be a number, where we pass it an array of numbers. Elsewhere in moodle I've seen this setup (get_in_or_equal followed by execute) be used with a foreach loop for each $param, which defeats the purpose of collecting them. There may be something I'm not getting here

Comment (dkleto) [reply cleaner/scheduled_tasks/classes/clean.php]: set_field_select might be what you're looking for here

File: cleaner/scheduled_tasks/classes/form/task_form.php [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+ */
+
+namespace cleaner_scheduled_tasks\form;
+use html_writer;
+use moodleform;
+
+defined('MOODLE_INTERNAL') || die();
+
+global $CFG;
+require_once("{$CFG->libdir}/formslib.php");
```
Comment (brendanheywood) [inline cleaner/scheduled_tasks/classes/form/task_form.php]: this looks odd, why the { } ?

Comment (kristian-94) [reply cleaner/scheduled_tasks/classes/form/task_form.php]: Just copied and pasted, probably from local/datacleaner/cleaner/muc/classes/form/upload_form.php that has the same format. There are a lot of places that format the require like that throughout moodle core as well though.

**Pull request merged**